README.md
<h1 align="center">
	<br>
	<img width="500" src="https://cdn.rawgit.com/leodido/material-palette/develop/media/logo.png" alt="Material Palette Generator">
	<br>
	<br>
	<br>
</h1>

[![CI](https://img.shields.io/travis/leodido/material-palette/master.svg?style=flat-square)][ci] [![NPM](https://img.shields.io/npm/v/material-palette.svg?style=flat-square)][npm] [![Coveralls branch](https://img.shields.io/coveralls/leodido/material-palette/master.svg?style=flat-square)][coveralls]

> **Generates material palettes for your colors**

Given an input color it calculates its palette approaching the logic of **[Google Material Design Color Palettes][mdpalettes]**.

**Since the algorithm behind their generation is not (yet) known this module approximates the colors values**.

It works everywhere: browsers, node, and so on.

## Install

With **[npm][npm]** do:

```
npm install material-palette
```

## API

This module exports a single function, the material palette generator.

### `materialpalette({ h: number, s: number, l: number })`

- It takes an object containing the h, s, and l fields

    - The field **h** must be a numeric into **[0, 360]**.

    - The fields **s** and **l** must be numerics into **[0, 100]**.

- It returns an object containing all the 14 palette variants (e.g., 50, 100, 200, ..., A700)

As always, more details **reading the [tests](test.js)** ...

## Example

Suppose you want to generate a material palette for **hsl(87, 100%, 22%)** color (i.e. **#3E7000**).

First of all import the generator.

```javascript
var materialpalette = require('materialpalette') // node
import materialpalette from 'materialpalette' // ES2015
window.materialpalette // browsers
```

Then use it.

```javascript
var ciao = materialpalette({ h: 87, s: 100, l: 22 })
```

Which results in the following object ...

```javascript
{
  '50': { h: 87, s: 100, l: 74 },
  '100': { h: 87, s: 100, l: 59 },
  '200': { h: 87, s: 100, l: 48 },
  '300': { h: 87, s: 100, l: 34 },
  '400': { h: 87, s: 100, l: 28 },
  '500': { h: 87, s: 100, l: 22 },
  '600': { h: 87, s: 100, l: 16 },
  '700': { h: 87, s: 100, l: 10 },
  '800': { h: 87, s: 100, l: 4 },
  '900': { h: 87, s: 100, l: 0 },
  'A100': { h: 92, s: 100, l: 46 },
  'A200': { h: 92, s: 100, l: 38 },
  'A400': { h: 92, s: 100, l: 21 },
  'A700': { h: 92, s: 100, l: 10 }
}
```

Matching, visually ...

![Material Colors Palette for #3E7000](palette-3e7000.png)

It's cool, isn't it?

## License

MIT © Leonardo Di Donato

---

[![Analytics](https://ga-beacon.appspot.com/UA-49657176-1/material-palette?flat)](https://github.com/igrigorik/ga-beacon)

[npm]:        https://npmjs.org/package/material-palette
[ci]:         https://travis-ci.org/leodido/material-palette
[coveralls]:  https://coveralls.io/r/leodido/material-palette?branch=master
[mdpalettes]: https://www.google.com/design/spec/style/color.html#color-color-palette


CHANGELOG.md
<a name="1.0.0"></a>
# 1.0.0 (2016-04-19)


### Build

* Code coverage support  ([0f194ec2e34884fc4dcc2348e0be7f7f9e6f34af](https://github.com/leodido/material-palette/commit/0f194ec2e34884fc4dcc2348e0be7f7f9e6f34af)), closes [#1](https://github.com/leodido/material-palette/issues/1)
* Configure rollupjs ([e6eaebc0828ed91cd05c7cb95f0aec926f202fa1](https://github.com/leodido/material-palette/commit/e6eaebc0828ed91cd05c7cb95f0aec926f202fa1))
* NPM ignore list ([77376cedc938f00325e384a758653dd57d0bb08a](https://github.com/leodido/material-palette/commit/77376cedc938f00325e384a758653dd57d0bb08a))
* NPM scripts ([deb634cc1ec753e33df3584c82ee3b1aff6e629a](https://github.com/leodido/material-palette/commit/deb634cc1ec753e33df3584c82ee3b1aff6e629a))

### Deps

* Setup ([b0a0830af2d898ac191d0800b464aa6ba058273a](https://github.com/leodido/material-palette/commit/b0a0830af2d898ac191d0800b464aa6ba058273a))

### Docs

* CHANGELOG ([0667b25b8c0513942394c269321c4bd7052f19f9](https://github.com/leodido/material-palette/commit/0667b25b8c0513942394c269321c4bd7052f19f9))
* License ([ecffad4d70e04f6abe96bf758cbbf8d075937df6](https://github.com/leodido/material-palette/commit/ecffad4d70e04f6abe96bf758cbbf8d075937df6))
* README ([45cd9cb1399b847afdcab6cb3c3af57312d4c59c](https://github.com/leodido/material-palette/commit/45cd9cb1399b847afdcab6cb3c3af57312d4c59c))

### New

* Automate changelog generation  ([d7646d56c1024ff76a56801e557d557b9b977a75](https://github.com/leodido/material-palette/commit/d7646d56c1024ff76a56801e557d557b9b977a75)), closes [#2](https://github.com/leodido/material-palette/issues/2)
* Dotfiles ([d5132ca3878d879839a6e795d5a90ae387f6d9ef](https://github.com/leodido/material-palette/commit/d5132ca3878d879839a6e795d5a90ae387f6d9ef))
* Material palette generator ([d152a08530304b0a3a56197fb98b87cf5e1eb153](https://github.com/leodido/material-palette/commit/d152a08530304b0a3a56197fb98b87cf5e1eb153))
* Tests ([3eebdeb6b719694da16fc59870326ba9af2de99b](https://github.com/leodido/material-palette/commit/3eebdeb6b719694da16fc59870326ba9af2de99b))
* Travis and coveralls integration  ([e730c89297e1d8f0da72bb7c6760b9636e2f221b](https://github.com/leodido/material-palette/commit/e730c89297e1d8f0da72bb7c6760b9636e2f221b)), closes [#3](https://github.com/leodido/material-palette/issues/3) [#4](https://github.com/leodido/material-palette/issues/4)

test.js
'use strict'

import test from 'ava'
import materialpalette from './'

test('should detect invalid input', t => {
    t.throws(() => materialpalette({ h: '36ABC', s: 100, l: 100 }), TypeError, 'Invalid input')
})

test('should handle (converting) numeric input strings', t => {
    t.plan(2)

    const palette = materialpalette({ h: '36', s: '100', l: '100' })
    t.not(palette, null)
    t.true(typeof palette === 'object')
})

test('should handle (ignoring) float input', t => {
    t.plan(2)

    const palette = materialpalette({ h: 36.3, s: 10.5, l: 99.9 })
    t.not(palette, null)
    t.true(typeof palette === 'object')
})

test('should detect invalid hue', t => {
    t.plan(2)

    t.throws(() => materialpalette({ h: 361, s: 100, l: 100 }), RangeError, `Hue must be an integer within [0, 360]; given 361`)
    t.throws(() => materialpalette({ h: -1, s: 100, l: 100 }), RangeError, `Hue must be an integer within [0, 360]; given -1`)
})

test('should detect invalid saturation', t => {
    t.plan(2)

    t.throws(() => materialpalette({ h: 36, s: 101, l: 100 }), RangeError, `Saturation must be an integer within [0, 100]; given 101`)
    t.throws(() => materialpalette({ h: 36, s: -1, l: 100 }), RangeError, `Saturation must be an integer within [0, 100]; given -1`)
})

test('should detect invalid lightness', t => {
    t.plan(2)

    t.throws(() => materialpalette({ h: 36, s: 100, l: 101 }), RangeError, `Lightness must be an integer within [0, 100]; given 101`)
    t.throws(() => materialpalette({ h: 36, s: 100, l: -1 }), RangeError, `Lightness must be an integer within [0, 100]; given 101`)
})

test('should accept limit values', t => {
    t.plan(4)

    let palette = materialpalette({ h: 0, s: 0, l: 0 })
    t.not(palette, null)
    t.true(typeof palette === 'object')

    palette = materialpalette({ h: 360, s: 100, l: 100 })
    t.not(palette, null)
    t.true(typeof palette === 'object')
})

test('should return an object of objects', t => {
    t.plan(170)

    const palette = materialpalette({ h: 359, s: 75, l: 75 })
    t.not(palette, null)
    t.true(typeof palette === 'object')

    for (let p of ['50', '100', '200', '300', '400', '500', '600', '700', '800', '900', 'A100', 'A200', 'A400', 'A700']) {
        t.true(palette.hasOwnProperty(p))
        t.not(palette[p], null)
        t.true(typeof palette[p] === 'object')
        for (let c of ['h', 's', 'l']) {
            t.true(palette[p].hasOwnProperty(c))
            t.true(Number.isInteger(palette[p][c]))
            t.true(palette[p][c] >= 0 && c === 'h' ? true : palette[p][c] <= 100)
        }
    }
})


index.m.js
'use strict'

/**
 * Minimize the maximum possible loss
 *
 * @param      {number}             val           The input value to test
 *
 * @return     {number}             A number between 0 and 100
 */
const minimax = val => Math.min(100, Math.max(0, val))

/**
 * Material Palette Generator
 *
 * It calculates all colors from base.
 * These colors were determined by finding all HSL values for each google palette.
 * Then calculating the differences in H, S, and L per color change individually.
 * Finally applying these here.
 *
 * @param      {Object}           param           The input color
 * @param      {number}           h               The hue ([0, 360]) of the input color
 * @param      {number}           s               The saturation ([0, 100]) of the input color
 * @param      {number}           l               The lightness ([0, 100]) of the input color
 *
 * @return     {Object}           Its palette
 * @property   {Object}           50              The variant 50 of the color
 * @property   {Object}           100             The variant 100 of the color
 * @property   {Object}           200             The variant 200 of the color
 * @property   {Object}           300             The variant 300 of the color
 * @property   {Object}           400             The variant 400 of the color
 * @property   {Object}           500             The input color
 * @property   {Object}           600             The variant 600 of the color
 * @property   {Object}           700             The variant 700 of the color
 * @property   {Object}           800             The variant 800 of the color
 * @property   {Object}           900             The variant 900 of the color
 * @property   {Object}           A100            The accent variant 100 of the color
 * @property   {Object}           A200            The accent variant 200 of the color
 * @property   {Object}           A400            The accent variant 400 of the color
 * @property   {Object}           A700            The accent variant 700 of the color
 */
const generator = ({ h, s, l }) => {
  h = Math.round(h)
  s = Math.round(s)
  l = Math.round(l)

  if (isNaN(h) || isNaN(s) || isNaN(l)) {
    throw new TypeError('Invalid input')
  }
  if (h < 0 || h > 360) {
    throw new RangeError(`Hue must be an integer within [0, 360]; given ${h}`)
  }
  if (s < 0 || s > 100) {
    throw new RangeError(`Saturation must be an integer within [0, 100]; given ${s}`)
  }
  if (l < 0 || l > 100) {
    throw new RangeError(`Lightness must be an integer within [0, 100]; given ${l}`)
  }

  return {
    '50': { h, s, l: minimax(l + 52) },
    '100': { h, s, l: minimax(l + 37) },
    '200': { h, s, l: minimax(l + 26) },
    '300': { h, s, l: minimax(l + 12) },
    '400': { h, s, l: minimax(l + 6) },
    '500': { h, s, l },
    '600': { h, s, l: minimax(l - 6) },
    '700': { h, s, l: minimax(l - 12) },
    '800': { h, s, l: minimax(l - 18) },
    '900': { h, s, l: minimax(l - 24) },
    'A100': { h: h + 5, s, l: minimax(l + 24) }, // { h, s, l: minimax(l + 52) }
    'A200': { h: h + 5, s, l: minimax(l + 16) }, // { h, s, l: minimax(l + 37) }
    'A400': { h: h + 5, s, l: minimax(l - 1) }, // { h, s, l: minimax(l + 6) }
    'A700': { h: h + 5, s, l: minimax(l - 12) } // { h, s, l: minimax(l - 12) }
  }
}

export default generator